applications/RomApplication/python_scripts/petrov_galerkin_training_utility.py

```diff
@@ -0,0 +1,158 @@
+
+        # Generate the matrix of residuals or projected Jacobians.
+        if self.basis_strategy=="Jacobian":
+            snapshots_matrix = self.__rom_residuals_utility.GetProjectedGlobalLHS()
+            if self.echo_level > 0 : KratosMultiphysics.Logger.PrintInfo("PetrovGalerkinTrainingUtility","Generated matrix of projected Jacobian.")
+        elif self.basis_strategy=="Residuals":
+            snapshots_matrix = []
+            files_to_read_and_delete = glob('*.res.mm')#TODO: Stop writing to disk.
+            for to_erase_file in files_to_read_and_delete:
+                non_converged_iteration_snapshot = KratosMultiphysics.Vector()
+                KratosMultiphysics.ReadMatrixMarketVector(to_erase_file, non_converged_iteration_snapshot)
+                snapshots_matrix.append(non_converged_iteration_snapshot)
+            snapshots_matrix = np.array(snapshots_matrix).T
+            if self.echo_level > 0 : KratosMultiphysics.Logger.PrintInfo("PetrovGalerkinTrainingUtility","Generating matrix of residuals.")
+        else:
+            err_msg = "\'self.basis_strategy\' is not available. Select either Jacobian or Residuals."
+            raise Exception(err_msg)
+
+        np_snapshots_matrix = np.array(snapshots_matrix, copy=False)
+        self.time_step_snapshots_matrix_container.append(np_snapshots_matrix)
+
+    def CalculateAndSaveBasis(self):
+        # Calculate the new basis and save
+        snapshots_basis = self.__CalculateResidualBasis()
+        self.__AppendNewBasisToRomParameters(snapshots_basis)
+
+
+    @classmethod
+    def __GetPetrovGalerkinTrainingDefaultSettings(cls):
+        default_settings = KratosMultiphysics.Parameters("""{
+                "train": false,
+                "basis_strategy": "Residuals",
+                "include_phi": false,
+                "svd_truncation_tolerance": 1.0e-6,
+                "echo_level": 0
+        }""")
+        return default_settings
+
+    def __CalculateResidualBasis(self):
+        # Set up the snapshots matrix for new basis
+        n_steps = len(self.time_step_snapshots_matrix_container)
+        snapshots_matrix = self.time_step_snapshots_matrix_container[0]
+        for i in range(1,n_steps):
+            del self.time_step_snapshots_matrix_container[0] # Avoid having two matrices, numpy does not concatenate references.
+            snapshots_matrix = np.c_[snapshots_matrix,self.time_step_snapshots_matrix_container[0]]
+        u_left,s_left,_,_ = RandomizedSingularValueDecomposition(COMPUTE_V=False).Calculate(
+            snapshots_matrix,
+            self.svd_truncation_tolerance)
+        del(snapshots_matrix)
+        #Include Phi in the span of Psi
+        if self.include_phi:
+            #Read Galerkin modes from RomParameters.json
+            u_right = self.__GetGalerkinBasis()
+            u , _= np.linalg.qr(np.c_[u_right,u_left])
```

Comment:
I'm not aware of any performance comparison but it might be interesting to try Eigen QR in here to see if it improves. Note that we now have interface to it through the `LinearSolversApplication`.